Icfg dot exporter

palaniappan1 opened the pull request:
The test does not seem to make sense, as I am comparing two strings that I know will be the same. I tried the resulting string in the https://dreampuf.github.io/GraphvizOnline/ website to visualize the ICFG. I also attach the resulting png image of the ICFG.
![graphviz](https://github.com/soot-oss/SootUp/assets/24876965/33be6160-ca8a-429d-8db3-fa029d3e4370)

palaniappan1 committed:
ICFG CallGraph export init

palaniappan1 committed:
ICFG DotExporter

palaniappan1 committed:
Merge remote-tracking branch 'origin/develop' into ICFG-DotExporter

palaniappan1 committed:
ICFG CallGraph Export

Test written does not make sense, as i am comparing two same strings.

palaniappan1 committed:
fmt commit

palaniappan1 committed:
Usage of Collection method changed

palaniappan1 committed:
Dummy Change for test

palaniappan1 committed:
fmt commit

kadirayk commented on sootup.analysis/src/main/java/sootup/analysis/interprocedural/icfg/JimpleBasedInterproceduralCFG.java:
```diff
@@ -148,6 +150,21 @@ public JimpleBasedInterproceduralCFG(
+  public String buildICFGGraph() {
+    ArrayList<StmtGraph> stmtGraphSet = new ArrayList<>();
+    // To Sort the methodSignature set with entrypoint as the first element.
+    Set<MethodSignature> methodSignatures = cg.getMethodSignatures();
+    methodSignatures.remove(mainMethodSignature);
+    Set<MethodSignature> modifiedMethodSignatures = new LinkedHashSet<>();
```
We don't need to get all the method signatures upfront. you can start from the main method and traverse the call graph using `cg.getCallsFrom()`. You can then add stmtGraphs to the dot graph as you traverse and add corresponding edges that represent the calls.

palaniappan1 replied:
CallGraph is not accessible in ICFGDotExporter.java, I need to add a dependency for accessing callgraph. So this is the plausible way of doing it I reckon. I can't use cg.callsFrom() because I need the hashcode of the invoke statement.

kadirayk replied:
What do you mean it's not accessible? You are calling `cg.getMethodSignatures()` ?

JonasKlauke commented on sootup.analysis/src/test/java/sootup/analysis/interprocedural/ifds/ICFGCallGraphTest.java:
```diff
@@ -0,0 +1,112 @@
+    JimpleBasedInterproceduralCFG icfg =
+        new JimpleBasedInterproceduralCFG(view, entryMethodSignature, false, false);
+    String expectedCallGraph =
+        "digraph G {\n"
```
Maybe instead of comparing the full result. you could check if all methods blocks are contained in the block and that they are connected by the calls.
Maybe you remove found parts from the string till the string is empty or you cant find a edge or code block in it.
this way it might be robust for small changes in order or something

JonasKlauke commented on sootup.analysis/src/test/resources/callgraph/ICFGExample.java:
```diff
@@ -0,0 +1,27 @@
+    public void entryPoint() {
+        String var1 = "SECRET";
+        String var2 = this.id(var1);
+        this.sink(var2);
```
test also a case where you can have multiple call targets for one method call. Like you call method m which can go to Superclass m and Subclass m

JonasKlauke replied:
but do this in another testcase.

palaniappan1 committed:
Changes in ICFGDotExporter

1. One more test case added
2.ICFGDotExporter moved from core to callgraph module

palaniappan1 committed:
import changed and fmt commit

kadirayk committed:
ICFGDotExporter

kadirayk committed:
fmt

palaniappan1 committed:
ICFG Changes

1. Created a folder ICFG in resources folder to add the files.
2. Converted the CallGraph String into Java Object for test

palaniappan1 committed:
Merge remote-tracking branch 'origin/develop' into ICFG-DotExporter

palaniappan1 committed:
Class files added

palaniappan1 committed:
ICFG Dot Exporter

kadirayk committed:
add test category

kadirayk committed:
fmt
